node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1552 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The bitfield distribution
+//!
+//! In case this node is a validator, gossips its own signed availability bitfield
+//! for a particular relay parent.
+//! Independently of that, gossips on received messages from peers to other interested peers.
+
+use codec::{Decode, Encode};
+use futures::{channel::oneshot, FutureExt};
+
+use keystore::KeyStorePtr;
+use sc_keystore as keystore;
+
+use node_primitives::{ProtocolId, View};
+
+use log::{trace, warn};
+use polkadot_erasure_coding::{
+	branch_hash, branches, obtain_chunks_v1 as obtain_chunks};
+use polkadot_primitives::v1::{
+	AvailableData, BlakeTwo256, CommittedCandidateReceipt, CoreState, ErasureChunk,
+	GlobalValidationData, Hash as H256, HashT, Id as ParaId, LocalValidationData,
+	OccupiedCoreAssumption, OmittedValidationData, PoV,
+	ValidatorId, ValidatorIndex, ValidatorPair,
+};
+use polkadot_subsystem::messages::*;
+use polkadot_subsystem::{
+	errors::{ChainApiError, RuntimeApiError}, ActiveLeavesUpdate, FromOverseer, OverseerSignal, SpawnedSubsystem,
+	Subsystem, SubsystemContext, SubsystemError,
+};
+use sc_network::ReputationChange as Rep;
+use std::collections::{HashMap, HashSet};
+use std::io;
+use std::iter;
+
+const TARGET: &'static str = "avad";
+
+#[derive(Debug, derive_more::From)]
+enum Error {
+	#[from]
+	Erasure(polkadot_erasure_coding::Error),
+	#[from]
+	Io(io::Error),
+	#[from]
+	Oneshot(oneshot::Canceled),
+	#[from]
+	Subsystem(SubsystemError),
+	#[from]
+	RuntimeApi(RuntimeApiError),
+	#[from]
+	ChainApi(ChainApiError),
+
+	Logic,
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+const COST_MERKLE_PROOF_INVALID: Rep = Rep::new(-100, "Merkle proof was invalid");
+const COST_NOT_A_LIVE_CANDIDATE: Rep = Rep::new(-51, "Candidate is not live");
+const COST_MESSAGE_NOT_DECODABLE: Rep = Rep::new(-100, "Message is not decodable");
+const COST_PEER_DUPLICATE_MESSAGE: Rep = Rep::new(-500, "Peer sent identical messages");
+const BENEFIT_VALID_MESSAGE_FIRST: Rep = Rep::new(15, "Valid message with new information");
+const BENEFIT_VALID_MESSAGE: Rep = Rep::new(10, "Valid message");
+
+/// Checked signed availability bitfield that is distributed
+/// to other peers.
+#[derive(Encode, Decode, Debug, Clone, PartialEq, Eq, Hash)]
+pub struct AvailabilityGossipMessage {
+	/// Ankor hash of the candidate this is associated to.
+	pub candidate_hash: H256,
```

Comment:
We haven't imported this type as `H256` elsewhere in the codebase

Reply:
Issues I've run into, was a myriad of `std::hash::Hash` and `Hash` collisions, I've seen the `H256` in `sp_core` i.e., I am open to suggestions.

Reply:
Our convention in Substrate is to import `std::hash::Hash` as `HashT`. We do this for other traits too, like `BlockT` and `HeaderT`.